generators/typescript/codegen/src/ast/core/Writer.ts

```diff
@@ -1,7 +1,81 @@
 import { AbstractWriter } from "@fern-api/generator-commons";
+import { Reference } from "../Reference";
+
+type ModuleName = string;
+
+export declare namespace Writer {
+    interface Args {
+        /* The name of the module that is being written to */
+        moduleName: string;
+    }
+}
 
 export class Writer extends AbstractWriter {
+    protected imports: Record<ModuleName, Reference[]> = {};
+    protected defaultImports: Record<ModuleName, Reference> = {};
+
+    /**
+     * Adds the given import under its module name.
+     */
+    public addImport(reference: Reference): void {
+        if (reference.module != null) {
+            if (reference.module.defaultExport ?? false) {
+                const existing = this.defaultImports[reference.module.moduleName];
+                if (existing == null) {
+                    this.defaultImports[reference.module.moduleName] = reference;
+                } else if (existing.name !== reference.name) {
+                    throw new Error(
+                        `Cannot have multiple default imports for module ${reference.module.moduleName}: ` +
+                            `got ${reference.name} but already had ${existing.name}`
+                    );
+                }
+            }
+            let moduleImports = this.imports[reference.module.moduleName];
+            if (moduleImports == null) {
+                this.imports[reference.module.moduleName] = [];
+                moduleImports = this.imports[reference.module.moduleName];
```

Comment:
nit: 

```
this.imports[reference.module.moduleName] ??= [];
moduleImports = this.imports[reference.module.moduleName];
```

You can skip the `null` check with the fancy `??=` operator 😅

Reply:
done